File: parser.php
<?php 
	
//--------------------------------настройки для работы с БД--------------------------------
	
	//Данные для подключения к БД
	
	
	include 'install/checkconf.php';
	
	$server = $MysqlIp;
	$username = $MysqlLogin;
	$password = $MysqlPassword;
	$database = $MysqlDatabase;
	$pathLog = $SquidLogfile;

	
	$mysqli = new mysqli( $server, $username, $password, $database ); //Устанавливаем соединение в базой мускула
	
	if ( $mysqli->connect_errno )
	{
		echo "Failed to connect to MySQL: (" . $mysqli->connect_errno . ") " . $mysqli->connect_error; //Не удалось установить соединение с базой мускула.
	}
	else
	{
		$mysqli->set_charset("utf8"); //Устанавливаем принудительно кодировку в UTF-8!
	}
//--------------------------------конец настройки--------------------------------
	echo "success connect to DB \n";
	
	
	
//----------------------------Получаем логины из БД--------------------------------
	$result = $mysqli->query( "SELECT login FROM users WHERE 1 " );
	$logins = $result->fetch_all( MYSQLI_NUM );

//--------------------------------закончили--------------------------------



	echo "start parse the logfile \n";
//--------------------------------Парсим лог --------------------------------

	$handler = fopen( $pathLog, "r" );

	$traffic = array();

	while( ( $temp = fgets( $handler ) ) !== false )
	{
		$parse = split("[ ]+", $temp ); //разбираем строку из файла лога на состовляющие

		for ( $i = 0; $i < count( $logins ); $i++ ) //для каждого логина в массиве проверяем текущую строку на совпадение логина
		{
			if ( strcasecmp($parse[7], $logins[$i][0]) == 0 )
			{
				array_push( $traffic, array( $parse[7], $parse[6], $parse[4], $parse[0] ) );
			}
		}
	}

	fclose( $handler );
//--------------------------------хватит парсить лог--------------------------------


	echo "parse data and add to DB \n";
//--------------------------------Обработка данных--------------------------------

	if ( count( $traffic ) > 0 )//если есть данные, разбираем на логин, сайт, байты, время и пишем в БД
	{
		$query = "SELECT dateTime FROM usersTraffic ORDER BY id DESC LIMIT 1";
		$result = $mysqli->query($query);
		$lastDate = $result->fetch_row();
		$lastDate = $lastDate[0]; //получаем дату последнего обновления траффика в базе

		$query = "INSERT INTO usersTraffic ( login, cite, bytes, dateTime ) VALUES "; //первая часть запроса, которая не изменится
		$firstQuery = $query;

		for ( $i = 0; $i < count( $traffic ); $i++ )
		{
			if ( $traffic[$i][3] > $lastDate ) //проверяем, чтобы траффик не повторялся
			{
				$query = $query."(\"".$traffic[$i][0]."\",\"".$traffic[$i][1]."\",\"".$traffic[$i][2]."\",\"".$traffic[$i][3]."\"),";
			}
		}
		if( $firstQuery != $query )
		{
			if ( substr($query, -1) == "," )
			{
				$query = substr($query, 0, -1);
			}
			$mysqli->query( $query );
		}

	}

	echo "count traffic for user \n";
//-------------------------------подсчитываем траффик за день для каждого пользователя----------------------------
	for( $i = 0; $i < count( $logins ); $i++ )
	{
		$query = $mysqli->query( "SELECT lastUpdate,trafficForDay FROM users WHERE login=\"".$logins[$i][0]."\"" );
		$result = $query->fetch_row();
		$lastUpdate = $result[0];
		$currentTraffic = $result[1];

		$query = $mysqli->query("SELECT SUM(bytes) FROM usersTraffic WHERE dateTime > \"".$lastUpdate."\" AND login =\"".$logins[$i][0]."\"");
		$result = $query->fetch_row();
		$traffic = round( $result[0]/1048576, 2 ) + $currentTraffic;

		if ( $traffic != 0 )
		{
			$query = $mysqli->query( "SELECT dateTime FROM usersTraffic WHERE login=\"".$logins[$i][0]."\" ORDER BY id DESC LIMIT 1" );
			$result = $query->fetch_row();
			$currentUpdate = $result[0];

			$mysqli->query("UPDATE users SET trafficForDay=\"".$traffic."\", lastUpdate= \"".$currentUpdate."\" WHERE login =\"".$logins[$i][0]."\"");
		}

//-----------Очищаем суточный траффик-------------
		$query = $mysqli->query( "UPDATE users SET trafficForDay=\"0\" WHERE CURTIME() >= \"23:58:00\" ");

	}
	echo "finish \n";
?>

File: parser.py
# -*- coding: utf-8

import mysql.connector
import os

server = ""
username = ""
password = ""
database = ""
pathLog = ""

try:
    dir = os.path.dirname(os.path.realpath(__file__))
    configFile = open(dir + "/install/config.cfg", "r")
    config = configFile.readlines()
    configFile.close()

    for line in config:
        if line.startswith("#") or line.startswith(" ") or line.startswith("\n"):
            continue
        key, value = line.split(" : ")
        if key == "MYSQL_ip":
            server = value.strip()
        elif key == "MYSQL_login":
            username = value.strip()
        elif key == "MYSQL_password":
            password = value.strip()
        elif key == "MYSQL_database":
            database = value.strip()
        elif key == "SQUID_logfile":
            pathLog = value.strip()

    if not server or not username or not password or not database or not pathLog:
        exit("Config file is incorrect")

    conn = mysql.connector.connect(user=username, password=password, host=server, database=database)
    cursor = conn.cursor()

except FileNotFoundError:
    exit("Config file not found")
except mysql.connector.Error:
    exit("Can't connect to mysql server or database")

def get_connection():
    global conn
    global cursor

    try:
        if not conn.is_connected():
            conn.reconnect()
            cursor = conn.cursor()
            return conn, cursor
        else:
            return conn, cursor
    except Exception as e:
        print(e)
        exit

query = "SELECT login FROM users WHERE 1"
cursor.execute(query)

login_array = cursor.fetchall()

query = "SELECT dateTime FROM usersTraffic ORDER BY id DESC LIMIT 1"
cursor.execute(query)

last_update = cursor.fetchone()
if last_update is None:
    last_update = 0.0
else:
    last_update = last_update[0]


def add_to_db( trafficArray ):
        query = "INSERT INTO usersTraffic ( login, site, bytes, dateTime ) VALUES "

        if len(trafficArray) < 1:
            return

        for traffic in trafficArray:
            user_login, user_url, user_bytes, user_dateTime = traffic

            if traffic == trafficArray[-1]:           
                query = query + "(\"" + user_login + "\",\"" + user_url + "\",\"" + user_bytes + "\",\"" + user_dateTime + "\")"
            else:
                query = query + "(\"" + user_login + "\",\"" + user_url + "\",\"" + user_bytes + "\",\"" + user_dateTime + "\"),"
        conn, cursor = get_connection()
        cursor.execute(query)
        conn.commit()

file = open(pathLog, 'r', encoding="utf-8", errors="ignore")

if os.path.exists('/tmp/sldap_poz.dat'):
    poz_file = open('/tmp/sldap_poz.dat', 'r+')
else:
    poz_file = open('/tmp/sldap_poz.dat', 'w+')

position = poz_file.readline()
if position == '':
    position = 0
else:
    position = int(position)

if position > 0:
    try:
       file.seek(position)
    except OSError:
        position = 0

trafficArray = []

for line in file:
    try:
        parsed_row = line.split()
        for login in login_array:
            if len(parsed_row) >= 7:
                if login[0].lower() == parsed_row[7].lower() and float(parsed_row[0]) > last_update:
                    traffic = []

                    traffic.append( parsed_row[7] )
                    traffic.append( parsed_row[6] )
                    traffic.append( parsed_row[4] )
                    traffic.append( parsed_row[0] )
                    trafficArray.append( traffic )
    except:
        continue

position = file.tell()
poz_file.seek(0)
poz_file.write(str(position))

poz_file.close()
file.close()

add_to_db( trafficArray )

query = "SELECT lastUpdate,trafficForDay,login FROM users WHERE login IN ("

for login in login_array:
    if login == login_array[-1]:
        query = query + "\"" + login[0] + "\")"
    else:
        query = query + "\"" + login[0] + "\","

conn, cursor = get_connection()
cursor.execute(query)

result = cursor.fetchall()

for row in result:
    last_update = row[0]
    current_traffic = row[1]
    login = row[2]
    query = "SELECT SUM(bytes) FROM usersTraffic WHERE dateTime > %s AND login = %s"
    conn, cursor = get_connection()
    cursor.execute(query, (last_update, login))
    row = cursor.fetchone()

    if row[0] is not None:
        traffic = float(round(row[0]/1048576, 2)) + current_traffic

        query = "SELECT dateTime FROM usersTraffic ORDER BY id DESC LIMIT 1"
        conn, cursor = get_connection()
        cursor.execute(query)
        last_update = cursor.fetchone()[0]

        if traffic > 0:
            print(login)
            query = "UPDATE users SET trafficForDay=%s, lastUpdate=%s WHERE login=%s"
            conn, cursor = get_connection()
            cursor.execute(query, (traffic, last_update, login))
            conn.commit()

query = "UPDATE users SET trafficForDay=\"0\" WHERE CURTIME() >= \"23:30:00\""
conn, cursor = get_connection()
cursor.execute(query)
conn.commit()

cursor.close()
conn.close()


